README.md
# Unity Ads Webview

## Requirements

- Node (`brew install node`) (Latest or at least Node 6+)
- Parallel (`brew install parallel`) for running things in parallel
- Python (`brew install python3`) for the local webserver
- Watchman (`brew install watchman`) for watching tests
- exiftool (`brew install exiftool`) for integration tests
- hub (`brew install hub`) for release scripts

### Optional

- segno (`pip install segno`) for qr code generation

## IDE

### WebStorm

1. Configure directories (Preferences -> Project: unity-ads-webview -> Directories)
	- Exclude
		1. .idea
		2. build
		3. node_modules
	- Tests
		1. test
		2. test-utils
2. Configure TypeScript (Preferences -> Languages & Frameworks -> TypeScript)
	1. Enable TypeScript
	2. Change compiler to use TypeScript installation from your project directory `node_modules/typescript/lib`
	3. Switch to use `tsconfig.json`
3. Enable TSLint (Preferences -> Languages & Frameworks -> TypeScript -> TSLint)
	1. Change TSLint package to use TSLint installation from your project directory `node_modules/tslint`

## Building

### Dependencies

- `make setup`

### Browser Build

Suggested testing browser is Google Chrome.

Build for the browser:

- `make build-browser`

Or to build continuously (on file changes), use:

- `make watch-browser`

To serve the build use: 

- `make start-server`

Navigate to: http://localhost:8000/build/browser/

### Development Build

- `make build-dev`

If you need to develop on legacy devices that do not support ES6, use the following:

- `MINIFY=1 make build-dev`

To build continuously (on file changes), use:

- `make watch-dev`

### Running development builds

- Change SDK configuration URL to point to local build (`http://LOCAL_IP:LOCAL_PORT/build/dev/config.json`)
  >**iOS:** change the return url of `getDefaultConfigUrl()` in [USRVSdkProperties.m](https://github.com/Applifier/unity-ads-ios/blob/master/UnityServices/Core/Properties/USRVSdkProperties.m)

  >**Android:** change the return url of `getDefaultConfigUrl()` in [SdkProperties.java](https://github.com/Applifier/unity-ads-android/blob/master/lib/src/main/java/com/unity3d/services/core/properties/SdkProperties.java)
- Change webview development config to point to local build (is done automagically by `make build-dev`) (`http://LOCAL_IP:LOCAL_PORT/build/dev/index.html`)
- Start local web server in project root (`make start-server`)

### Adding new icon
 **Add icon to existing font file**
 - Copy the svg file into an existing src/fonts/* folder
 - `make build-fonts`

 **Create new font file**
 - Create a new folder under src/fonts and copy the svg file(s) into the folder
 - Modify tools/generate_fonts.js to include the newly added folder
 - `make build-fonts`

`make build-fonts` will generate svg, ttf, and woff files in src/fonts/*/generated and styl file in src/styl/generated

## Testing

### Local tests

- `make test` for both unit and integration tests
- `make test-unit`
- `make test-integration`
- `npm run test` for jest tests

To watch tests, use:

- `make watch-test`
- `TEST_FILTER=ObservableTest make watch-test`

To debug tests (opens a browser running the tests and pauses it immediately), use:

- `DEBUG=1 make test`

To filter what tests are being run, use:

- `TEST_FILTER=<RegExp> make test`

Or combine both:

- `DEBUG=1 TEST_FILTER=ObservableTest make test-unit`

### Test coverage

- `make test-coverage` will generate a coverage report at `build/coverage/index.html`

### Hybrid tests

- Change SDK configuration URL to point to local build (`http://LOCAL_IP:LOCAL_PORT/build/test/config.json`)
- Change webview test config (src/test-config.json) to point to local build (`http://LOCAL_IP:LOCAL_PORT/build/test/index.html`)
- `make build-test`
- Run hybrid test suite from the SDK

#### Automatic builds for Android

https://jenkins.prd.mz.internal.unity3d.com/job/Applifier/job/unity-ads-sdk-tests/job/ads-sdk-hybrid-test-android/

#### Automatic builds for iOS

https://jenkins.prd.mz.internal.unity3d.com/job/Applifier/job/unity-ads-sdk-tests/job/ads-sdk-hybrid-test-ios/

### Browser build tests

- `make build-browser start-server`
- `node test-utils/headless.js`

### Deployment tests

#### Android
https://jenkins.prd.mz.internal.unity3d.com/job/Applifier/job/unity-ads-sdk-tests/job/ads-sdk-systest-android/

#### iOS
https://jenkins.prd.mz.internal.unity3d.com/job/Applifier/job/unity-ads-sdk-tests/job/ads-sdk-systest-ios/



CHANGELOG.md
# Wed Aug 19 15:45:36 UTC 2020

* Add Identifier for Install (IDFI) via new InstallInfo class. [#16977](https://github.com/Applifier/unity-ads-webview/pull/16977)
* [Load] Scale Load/Fill Events to 55% [#17143](https://github.com/Applifier/unity-ads-webview/pull/17143)

# Tue Aug 18 21:54:28 UTC 2020

* [Load] Scale Load/Fill Events to 40% [#17108](https://github.com/Applifier/unity-ads-webview/pull/17108)
* [Signal Mapping] Add Device Free Space to Config Request [#17110](https://github.com/Applifier/unity-ads-webview/pull/17110)

# Tue Aug 18 11:52:10 UTC 2020

* Enabled Video + Playable experiment [#16982](https://github.com/Applifier/unity-ads-webview/pull/16982)

# Tue Aug 18 10:23:19 UTC 2020

* [Load V5] Enable Auction V6 protocol [#17045](https://github.com/Applifier/unity-ads-webview/pull/17045)
* [Load] Ramp Load/Fill Events to 25% [#17050](https://github.com/Applifier/unity-ads-webview/pull/17050)
* Add Video + Playable Format Support [#17046](https://github.com/Applifier/unity-ads-webview/pull/17046)

# Mon Aug 17 22:24:28 UTC 2020

* [LoadV5]Indexed encrypted data in request body [#16791](https://github.com/Applifier/unity-ads-webview/pull/16791)
* Added endScreenSettings in preparation for the Video + Playable initiative [#16984](https://github.com/Applifier/unity-ads-webview/pull/16984)
* [Load V5] Auction Protocol V6 support [#16979](https://github.com/Applifier/unity-ads-webview/pull/16979)

# Fri Aug 14 23:00:50 UTC 2020

* Add missing third quartile events in PTS [#16978](https://github.com/Applifier/unity-ads-webview/pull/16978)
* Add click sent/received metrics for MRAID webplayer [#16792](https://github.com/Applifier/unity-ads-webview/pull/16792)
* [Signal Mapping] Add Device Fields to Config Request [#16888](https://github.com/Applifier/unity-ads-webview/pull/16888)
* [Load] Utilize LoadAndFillEventManager [#16556](https://github.com/Applifier/unity-ads-webview/pull/16556)
* [Load] Sample Load/Fill Events at 10% [#16856](https://github.com/Applifier/unity-ads-webview/pull/16856)

# Tue Aug 11 23:51:16 UTC 2020

* [Load V5] Revert Reverse AB test to Normal AB Test [#16887](https://github.com/Applifier/unity-ads-webview/pull/16887)

# Tue Aug 11 12:37:02 UTC 2020

* Support new useUnityAttDialog config field [#16822](https://github.com/Applifier/unity-ads-webview/pull/16822)

# Mon 10 Aug 2020 22:14:28 UTC

* [EES] Remove default route for the external end screen [#16886](https://github.com/Applifier/unity-ads-webview/pull/16886)
* [Chore] Remove load called counter [#16662](https://github.com/Applifier/unity-ads-webview/pull/16662)
* [Load V5] Revert placement grouping experiment [#16855](https://github.com/Applifier/unity-ads-webview/pull/16855)
* [Chore] Send frequency cap as separate metric [#16854](https://github.com/Applifier/unity-ads-webview/pull/16854)
* [Open Measurement] Refactor AdViewBuilder PR1: Utilize screen.width and screen.height  [#15533](https://github.com/Applifier/unity-ads-webview/pull/15533)
* Revert Testing PR's [#16825](https://github.com/Applifier/unity-ads-webview/pull/16825)

# Fri Aug  7 19:26:43 UTC 2020

* [Load V5] Add support for frequency capping [#16695](https://github.com/Applifier/unity-ads-webview/pull/16695)
* MDP-225 Bump up optout event to v4 [#16761](https://github.com/Applifier/unity-ads-webview/pull/16761)

# Pending

* [3.4.8] Merge to Master [#16793](https://github.com/Applifier/unity-ads-webview/pull/16793)
* [3.4.8] Add TrackingManagerApi [#16223](https://github.com/Applifier/unity-ads-webview/pull/16223)
* [3.4.8] Add support for new iOS 14 device info fields [#16432](https://github.com/Applifier/unity-ads-webview/pull/16432)
* [3.4.8] Test Logs [#16433](https://github.com/Applifier/unity-ads-webview/pull/16433)
* [3.4.8] Add iOS 14 Fields to iOSDeviceInfo Model [#16567](https://github.com/Applifier/unity-ads-webview/pull/16567)
* [3.4.8] Add SKAdNetwork API methods [#16527](https://github.com/Applifier/unity-ads-webview/pull/16527)
* [3.4.8] Add New iOS Fields to Configuration Request [#16693](https://github.com/Applifier/unity-ads-webview/pull/16693)
* [3.4.8] Add iOS fields to Auction Request and Operative Events [#16697](https://github.com/Applifier/unity-ads-webview/pull/16697)
* [3.4.8] Add Logging for App Tracking Transparency for Testing [#16563](https://github.com/Applifier/unity-ads-webview/pull/16563)

# Thu Aug  6 11:24:22 UTC 2020

* Added check for EndScreenType inside PerformanceAdUnitParametersFactory [#16729](https://github.com/Applifier/unity-ads-webview/pull/16729)
* Disable Privacy SDK A/B test [#16700](https://github.com/Applifier/unity-ads-webview/pull/16700)

# Wed Aug  5 22:26:46 UTC 2020

* Adding metric events for reported length vs actual length of videos [#16598](https://github.com/Applifier/unity-ads-webview/pull/16598)

# Tue  4 Aug 2020 21:09:26 UTC

* [EES] Support for metrics coming from iframe [#16698](https://github.com/Applifier/unity-ads-webview/pull/16698)
* Rename unityads-logo to icon-unity-ads-logo for clarity [#16699](https://github.com/Applifier/unity-ads-webview/pull/16699)
* [AUI] Fix a bug with the Shining button animation on iOS [#16692](https://github.com/Applifier/unity-ads-webview/pull/16692)
* Add more creative IDs to NestedIframePlayableCreatives [#16696](https://github.com/Applifier/unity-ads-webview/pull/16696)

# Fri Jul 31 15:52:44 UTC 2020

* [AUI] Add TiltedEndScreen to AutomatedExperiment [#16566](https://github.com/Applifier/unity-ads-webview/pull/16566)
* Improve metrics and iframe loading [#16603](https://github.com/Applifier/unity-ads-webview/pull/16603)

# Fri Jul 31 08:14:05 UTC 2020

* [Latency] Remove Dynamic CacheMode Experiment and Blocklist [#16601](https://github.com/Applifier/unity-ads-webview/pull/16601)
* Added endScreenType to CometCampaignParser [#16526](https://github.com/Applifier/unity-ads-webview/pull/16526)
* [AUI] Default to green when color matching is not supported [#16557](https://github.com/Applifier/unity-ads-webview/pull/16557)

tools/purge.js
/* eslint-disable @typescript-eslint/no-var-requires */
const fetch = require('node-fetch');
const crypto = require('crypto');
const querystring = require('querystring');
const childProcess = require('child_process');
const releases = require('./releases');

const cdnConfig = {
    'akamai': {
        'base_urls': [
            'cdn-akamai.unityads.unity3d.com',
            'cdn.unityads.unity3d.com',
            'config.unityads.unity3d.com',
            'webview.unityads.unity3d.com',
            'config-cn.unityads.unity3d.com'
        ],
        'host': process.env.AKAMAI_HOST,
        'access_token': process.env.AKAMAI_ACCESS_TOKEN,
        'client_secret': process.env.AKAMAI_CLIENT_SECRET,
        'client_token': process.env.AKAMAI_CLIENT_TOKEN
    },
    'highwinds': {
        'base_urls': [
            'cdn-highwinds.unityads.unity3d.com',
            'cdn.unityads.unity3d.com',
            'geocdn.unityads.unity3d.com',
            'config.unityads.unity3d.com',
            'webview.unityads.unity3d.com'
        ],
        'check_url': 'cdn-highwinds.unityads.unity3d.com',
        'bearer': process.env.HIGHWINDS_BEARER,
        'account_hash': process.env.HIGHWINDS_ACCOUNT_HASH
    },
    'alibabacloud': {
        'base_urls': [
            'china-cdn2.unityads.unity3d.com',
            'cdn.unityads.unity3d.com',
            'geocdn.unityads.unity3d.com',
            'config.unityads.unity3d.com',
            'webview.unityads.unity3d.com',
            'config.unityads.unitychina.cn',
            'webview.unityads.unitychina.cn'
        ],
        'check_url': 'china-cdn2.unityads.unity3d.com',
        'access_key_id': process.env.ALIBABACLOUD_ACCESS_KEY_ID,
        'access_key_secret': process.env.ALIBABACLOUD_ACCESS_KEY_SECRET
    },
    'tencentcloud': {
        'base_urls': [
            'china-cdn3.unityads.unitychina.cn',
            'config.unityads.unitychina.cn',
            'webview.unityads.unitychina.cn'
        ],
        'check_url': 'china-cdn3.unityads.unitychina.cn',
        'secret_id': process.env.TENCENT_SECRET_ID,
        'secret_key': process.env.TENCENT_SECRET_KEY
    }
};

let branch = process.env.TRAVIS_BRANCH;
if (!branch) {
    throw new Error('Invalid branch: ' + branch);
}

const branchList = releases.getNativeVersionsSupportedByBranch(branch);

const commit = process.env.TRAVIS_COMMIT;
if (!commit) {
    throw new Error('Invalid commit: ' + commit);
}

const paths = [
    '/release/config.json',
    '/test/config.json'
];

const urlsFromPath = (urlRoot, baseUrls, path, addHttps) => {
    return baseUrls.map((baseUrl) => {
        return (addHttps ? 'https://' : '') + baseUrl + urlRoot + path;
    });
};

const flatten = (array) => {
    return array.reduce((flatArray, array) => {
        return flatArray.concat(Array.isArray(array) ? flatten(array) : array);
    }, []);
};

const fetchRetry = (url, options, retries, delay) => {
    const doFetch = () => {
        return fetch(url, options ? options : {}).catch(error => {
            console.dir(error);
            console.log('Retrying in ' + delay + 'ms');
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                    fetchRetry(url, options, --retries, delay)
                        .then((response) => resolve(response))
                        .catch((error) => reject(error));
                }, delay);
            });
        });
    };
    if (retries >= 0) {
        return doFetch();
    }
    return Promise.reject('Failed to fetch "' + url + '" after retries');
};

const checkConfigJson = (url, version) => {
    console.log('Checking "' + url + '"');
    const doFetch = () => {
        return fetchRetry(url, {}, 5, 5000).then(res => res.json()).then(configJson => {
            if (configJson.version !== version) {
                console.log('Invalid version "' + configJson.version + '" from "' + url + '"');
                return new Promise((resolve) => {
                    setTimeout(() => {
                        doFetch().then(() => resolve());
                    }, 5000);
                });
            }
        });
    };
    return doFetch();
};

let purgeAkamai = (urlRoot) => {
    let urls = flatten(paths.map(path => {
        return urlsFromPath(urlRoot, cdnConfig.akamai.base_urls, path, true);
    }));

    const executeShell = (command) => {
        childProcess.execSync(command, (err, stdout, stderr) => {
            console.log(stdout);
            console.log(stderr);
            if (stdout.indexOf('[OK]') === -1 || stderr.indexOf('API Error') > -1) {
                throw new Error(`Akamai purge request failed`);
            }
        });
    };

    console.log('Starting Akamai purge of: ');
    console.dir(urls);
    const edgerc = `[ccu]\nhost = ${cdnConfig.akamai.host}\nclient_token = ${cdnConfig.akamai.client_token}\nclient_secret = ${cdnConfig.akamai.client_secret}=\naccess_token = ${cdnConfig.akamai.access_token}`;
    const createEdgercFile = `echo $'${edgerc}' > $(pwd)/.edgerc`;
    const purgeAkamai = `akamai purge --edgerc $(pwd)/.edgerc invalidate ${urls.join(' ')}`;
    executeShell(`${createEdgercFile} && ${purgeAkamai}`);
};

let purgeHighwinds = (urlRoot) => {
    let urls = flatten(paths.map(path => {
        return urlsFromPath(urlRoot, cdnConfig.highwinds.base_urls, path, true);
    }));

    console.log('Starting Highwinds purge of: ');
    console.dir(urls);

    let body = {
        'list': urls.map((url) => {
            return { 'url': url };
        })
    };

    const endpoint = 'https://striketracker.highwinds.com/api/v1/accounts/' + cdnConfig.highwinds.account_hash + '/purge';

    return fetchRetry(endpoint, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + cdnConfig.highwinds.bearer
        },
        body: JSON.stringify(body)
    }, 5, 5000).then(res => {
        if (res.status !== 200) {
            console.log('Highwinds purge request failed, HTTP ' + res.status);
            return res.text().then(body => {
                console.dir(body);
                throw new Error('Highwinds purge request failed ' + res.status + ': ' + body);
            });
        }
        return res.text();
    }).then(body => {
        console.dir(body);
        console.log('Highwinds purge request successful');
        return Promise.all(paths.map(path => checkConfigJson('https://' + cdnConfig.highwinds.check_url + urlRoot + path, commit)));
    });
};

let purgeAliBabaCloud = (urlRoot) => {
    let urls = flatten(paths.map(path => {
        return urlsFromPath(urlRoot, cdnConfig.alibabacloud.base_urls, path, false);
    }));

    console.log('Starting AliBabaCloud purge of: ');
    console.dir(urls);

    const secret = cdnConfig.alibabacloud.access_key_secret;

    const getParameters = (path) => {
        return {
            Format: 'JSON',
            Version: '2014-11-11',
            AccessKeyId: cdnConfig.alibabacloud.access_key_id,
            SignatureMethod: 'HMAC-SHA1',
            Timestamp: encodeURIComponent((new Date()).toISOString().slice(0, -5) + 'Z'),
            SignatureVersion: '1.0',
            SignatureNonce: Math.random() * Math.pow(10, 18),
            Action: 'RefreshObjectCaches',
            ObjectPath: encodeURIComponent(path)
        };
    };

    const getCanonicalized = (parameters) => {
        let canonicalized = [];
        Object.keys(parameters).sort().forEach((key) => {
            if (parameters[key]) {
                canonicalized.push(encodeURIComponent(key) + '=' + encodeURIComponent(parameters[key]));
            }
        });
        return canonicalized.join('%26');
    };

    const getStringToSign = (value) => {
        return 'GET&%2F&' + value.replace(/=/g, '%3D');
    };

    const getSignedString = (value) => {
        const hmac = crypto.createHmac('sha1', secret + '&');
        hmac.update(value);
        return hmac.digest('base64');
    };

    const getUrls = (urls) => {
        return urls.map(url => {
            const parameters = getParameters(url);
            const signature = getSignedString(getStringToSign(getCanonicalized(parameters)));
            parameters.Signature = encodeURIComponent(signature);
            return 'https://cdn.aliyuncs.com/?' + Object.entries(parameters).map(([key, value]) => {
                return key + '=' + value;
            }).join('&');
        });
    };

    return Promise.all(getUrls(urls).map((url) => {
        return fetchRetry(url, {}, 5, 5000).then(res => {
            if (res.status !== 200) {
                throw new Error(`AliBabaCloud purge request failed with HTTP code: ${res.status}`);
            }
            return res.text();
        }).then(body => {
            console.dir(body);
        });
    })).then(() => {
        console.log('AliBabaCloud purge request successful');
        return Promise.all(paths.map(path => checkConfigJson('https://' + cdnConfig.alibabacloud.check_url + urlRoot + path, commit)));
    });
};

let purgeTencent = (urlRoot) => {

    let urls = flatten(paths.map(path => {
        return urlsFromPath(urlRoot, cdnConfig.tencentcloud.base_urls, path, true);
    }));

    console.log('Starting Tencent purge of: ');
    console.dir(urls);

    let paramObject = {
        Action: 'PurgeUrlsCache',
        Nonce: Math.random() * Math.pow(10, 18),
        Region: '',
        SecretId: cdnConfig.tencentcloud.secret_id,
        Timestamp: Math.floor(new Date().getTime() /1000),
        Token: '',
        Version: '2018-06-06'
    };

    const setupUrls = (urls, paramObject) => {
        urls.map((url, index) => {
            paramObject['Urls.' + index] = url;
        });
    };

    const getCanonicalized = (params, urlEncodeValue) => {
        const canonicalized = Object.entries(params).sort().map(([key, value]) => {
            const val = urlEncodeValue ? encodeURIComponent(value) : value;
            return key + '=' + val;
        }).join('&');
        return canonicalized;
    };

    const getStringToSign = (params) => {
        const signatureSrc = 'GETcdn.tencentcloudapi.com/?' + getCanonicalized(params, false);
        return signatureSrc;
    };

    const getSignedString = (signatureSrc, secretKey) => {
        const signature = crypto.createHmac('sha1', secretKey).update(signatureSrc).digest('base64');
        return signature;
    };

    const getRequestUrl = () => {

        setupUrls(urls, paramObject);
        const signSrc = getStringToSign(paramObject);
        const signature = getSignedString(signSrc, cdnConfig.tencentcloud.secret_key);

        paramObject['Signature'] = signature;
        const urlCandidate = getCanonicalized(paramObject, true);
        const signedUrl = 'https://cdn.tencentcloudapi.com/?' + urlCandidate;
        return signedUrl;
    };

    return fetchRetry(getRequestUrl(), {}, 5, 5000).then(res => {
        if (res.status !== 200) {
            throw new Error(`Tencent CDN purge request failed with HTTP code: ${res.status}`);
        }
        return res.text();
    }).then(body => {
        console.dir(body);
        console.log('Tencent CDN purge request successful');
        return Promise.all(paths.map(path => checkConfigJson('https://' + cdnConfig.tencentcloud.check_url + urlRoot + path, commit)));
    });
};

let purgeList = [];

branchList.forEach((branch) => {
    purgeList.push(
        purgeAkamai('/webview/' + branch),
        purgeHighwinds('/webview/' + branch),
        purgeAliBabaCloud('/webview/' + branch),
        purgeTencent('/webview/' + branch)
    );
});

Promise.all(purgeList).then(() => {
    console.log('Successfully purged all CDNs!');
}).catch(error => {
    console.dir(error);
    process.exit(1);
});
